### ISSUE TITLE
Add support for OpenMP and disabling parallelisation for Eigen

### ISSUE THREAD

**Issue by daljit46** [MRtrix3/mrtrix3#2949]
The Python configure script had support for enabling the use of OpenMP to enable multithreading in Eigen. If support for OpenMP is disabled, the project was built with `EIGEN_DONT_PARALLELIZE` (see [here](https://eigen.tuxfamily.org/dox/TopicMultiThreading.html). This functionality hasn't been ported to the CMake scripts, we should probably add a new option `MRTRIX_USE_OPENMP` to address this.

**Comment by Lestropie** [comment]
It may be preferable for control of this mechanism to be more fine-grained across the project. For anything where MRtrix is itself utilising extensive multi-threading, it will likely be optimal to disable Eigen multi-threading outright. However I'm currently utilising existing MRtrix code previously written for smaller matrix operations running just once on much larger matrices, and OpenMP-ing only there would be highly beneficial.

I wonder if the best approach is:
1.  Do not explicitly disable Eigen OpenMP unless instructed to do so at configuration time (or OpenMP is unavailable)
2.  Run `Eigen::setNbThreads(1)` in `int main()`
3.  Developers can call `Eigen::setNbThreads(Thread::number_of_threads())` at their own discretion

?

**Comment by Lestropie** [comment]
Note on above: Eigen strongly recommends against using the number of available threads for dense matrix manipulations, instead advocating for specifying the number of physical cores. This is however nontrivial to determine, and would likely require separate implementations for all three supported OS's. Python's `psutil` is probably the best reference.

**Comment by daljit46** [comment]
> Note on above: Eigen strongly recommends against using the number of available threads for dense matrix manipulations, instead advocating for specifying the number of physical cores.

Hmm, wouldn't `std::thread::hardware_concurrency` be suitable? We are using this [internally](https://github.com/MRtrix3/mrtrix3/blob/35dc21e2e0b91b40f352830ee16d847e47a872e5/core/thread.cpp#L74) already.

**Comment by Lestropie** [comment]
`std::thread::hardware_concurrency` reports twice the number of physical cores if hyperthreading is enabled. Even though "hardware" is in the name, it's not the physical number of cores. 

In MRtrix3 we get measurable performance benefits for executing this number of threads over manually specifying the number of physical cores, because for most computational tasks threads often spend time waiting for data to get from RAM to cache. But the Eigen documentation suggests that their optimisation is so highly targeted that executing more threads than there are physical cores is detrimental.

**Comment by daljit46** [comment]
OK, interesting. I don't think it'd be difficult to add a class to query physical cores and/or logical cores in C++. However, if we do want to take this approach, I think we should at least test whether what Eigen suggests is actually true in practice and check if using `std::thread::hardware_concurrency` performs worse.

**Comment by Lestropie** [comment]
Agreed. I reckon priority should be to just get Eigen multithreading working in isolated scenarios. Once that's in place, I can do a manual override of thread count in the context in which I'm currently operating. If halving thread count for Eigen has a noticeable benefit in a command that's slated for future addition, only then would it justify the effort.
